ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/Tests/Rollback.hs

```diff
@@ -0,0 +1,85 @@
+
+import           Control.Monad.IOSim (runSimOrThrow)
+import           Data.Maybe (fromJust)
+import           Ouroboros.Consensus.Block (ChainHash (..))
+import           Ouroboros.Consensus.Config.SecurityParam
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           Test.Consensus.BlockTree (BlockTree (..), BlockTreeBranch (..))
+import           Test.Consensus.Genesis.Setup
+import           Test.Consensus.PeerSimulator.Run (noTimeoutsSchedulerConfig)
+import           Test.Consensus.PeerSimulator.StateView
+import           Test.Consensus.PointSchedule
+import qualified Test.QuickCheck as QC
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock (TestBlock, unTestHash)
+
+tests :: TestTree
+tests = testGroup "rollback" [
+  testProperty "can rollback" (prop_rollback True),
+  testProperty "cannot rollback" (prop_rollback False)
+  ]
+
+-- | The boolean parameter specifies whether we are interested in the property
+-- when we can rollback.
+prop_rollback :: Bool -> QC.Gen QC.Property
+prop_rollback wantRollback = do
+  genesisTest <- genChains 1
+
+  let schedule = rollbackSchedule (gtBlockTree genesisTest)
+
+  -- | We consider the test case interesting if we want a rollback and we can
+  -- actually get one, or if we want no rollback and we cannot actually get one.
+  pure $
+    wantRollback == canRollbackFromTrunk (gtSecurityParam genesisTest) (gtBlockTree genesisTest)
+    ==>
+      runSimOrThrow $ runTest schedulerConfig genesisTest schedule $ \StateView{svSelectedChain} ->
+        let headOnAlternativeChain = case AF.headHash svSelectedChain of
+              GenesisHash    -> False
+              BlockHash hash -> any (0 /=) $ unTestHash hash
+        in
+        -- The test passes if we want a rollback and we actually end up on the
+        -- alternative chain or if we want no rollback and end up on the trunk.
+        wantRollback == headOnAlternativeChain
+
+  where
+    schedulerConfig = noTimeoutsSchedulerConfig defaultPointScheduleConfig
+
+    -- A schedule that advertises all the points of the trunk, then switches to
+    -- the first alternative chain of the given block tree.
+    --
+    -- PRECONDITION: Block tree with at least one alternative chain.
+    rollbackSchedule :: BlockTree TestBlock -> PointSchedule
+    rollbackSchedule blockTree =
+      let trunk = btTrunk blockTree
+          branch = btbSuffix $ head $ btBranches blockTree
+          states = banalStates trunk ++ banalStates branch
+          peers = peersOnlyHonest states
+          pointSchedule = balanced peers
+       in fromJust pointSchedule
+
+    -- | Whether it is possible to roll back from the trunk after having served
+    -- it fully, that is whether there is an alternative chain that forks of the
+    -- trunk less than 'k' blocks from the tip and that is longer than the
+    -- trunk.
```

Comment:
This is just as true for the part about `k`. The important property is indeed not about the tip but what last point you served: it has to be no more than `k` from the intersection and there has to be more blocks to serve on the alternative branch. For now, we only serve the tip, though it could be an option to serve less than this. Would you want me to make this clearer here?

Reply:
Upon rereading the documentation, it looks sufficiently clear to me now. I still think that the name of the function could be made more precise.

Reply:
Fine by me! I will take care of this in an hour or so.

Reply:
Done in f9a9df975.